rust_src/src/frames.rs

```diff
@@ -1,14 +1,57 @@
 //! Generic frame functions.
 
+use std::ptr;
+
 use remacs_macros::lisp_fn;
-use remacs_sys::Lisp_Frame;
-use remacs_sys::selected_frame as current_frame;
+use remacs_sys::{selected_frame as current_frame, Lisp_Frame};
+use remacs_sys::Qframe_live_p;
+use remacs_sys::fget_terminal;
 
 use lisp::{ExternalPtr, LispObject};
 use lisp::defsubr;
 
 pub type LispFrameRef = ExternalPtr<Lisp_Frame>;
 
+impl LispFrameRef {
+    #[inline]
+    pub fn is_live(&self) -> bool {
+        unsafe { fget_terminal(self.as_ptr()) != ptr::null() }
+    }
+}
+
+pub fn frame_or_selected(object: LispObject) -> LispFrameRef {
```

Comment:
Add some docstrings?

Reply:
What could I say that is not obvious from the code? if nil return selected frame else frame